### PR TITLE
Make qsort_r argument order the same via macros

### LINKED ISSUE #7586: gmt surface gave segfault with pixel-node registration but not gridline-node registration, only on M1 architecture
Running a tutorial this week, we discovered 4 people having the same issue:  ```gmt surface``` gave different behaviors for ```-rg``` (success)  and ```-rp``` (segfault).  They were all M1 macs.  No issues are reported on linux or Intel-chip macs for the same commands and same data; and strangely, a few M1 macs also succeeded.  The data file we used is attached.  The problem might be instigated by something in the data file, but still, the unexpected failure of one registration over the other, and apparently just on M1 architecture, feels like it's a bug nonetheless.    Please let me know if you need help reproducing it and I'll rope in some of the participants who encountered it. 

```gmt surface xy_vert.txt -Gvert.grd -R-125.8/-122.6/39.3/41.7 -I0.0064 -rg```
--> This worked perfectly. 

```gmt surface xy_vert.txt -Gvert.grd -R-125.8/-122.6/39.3/41.7 -I0.0064 -rp```
--> This produced error message: 

ERROR: Caught signal number 11 (Segmentation fault: 11) at
0 libgmt.6.dylib 0x0000000102a12f64 _quicksort + 1528
zsh: trace trap gmt surface xy_vert.txt -Gvert.grd -R-125.8/-122.6/39.3/41.7 -I0.0064 -rp

**System information**
As one example of a system that encountered this issue: 
System: 2021 Macbook Pro, OS Ventura 13.4, Apple M1 Pro
arch: arm64
GMT: 6.4.0

[xy_vert.txt](https://github.com/GenericMappingTools/gmt/files/11919096/xy_vert.txt)

### PR DESCRIPTION
**Pull request by PaulWessel**
THis is trying to fix #7586. For background, see https://stackoverflow.com/questions/39560773/different-declarations-of-qsort-r-on-mac-and-linux.

Not doing it right yet but we will have to deal with different argument order, since Linux has a different order than macOS.  Here is what my attempt of a macro that is in effect on Linus:

`#define QSORT_R(base, nel, width, compar, thunk) qsort_r(base, nel, width, compar, thunk);`

while if Apple the we have 

`#define QSORT_R(base, nel, width, compar, thunk) qsort_r(base, nel, width, thunk, compar);`

All places were we used to call qsort_r now calls Q_SORT_R instead to get the order right.  However, on Apple is still get compile errors like this:

```
[28/231] Building C object src/CMakeFiles/gmtlib.dir/gmt_support.c.o
/Users/pwessel/UH/RESEARCH/CVSPROJECTS/GMTdev/gmt-dev/src/gmt_support.c:14845:48: warning: incompatible function pointer types passing 'int (const void *, const void *, void *)' to parameter of type 'int (* _Nonnull)(void *, const void *, const void *)' [-Wincompatible-function-pointer-types]
        QSORT_R (L, nl, sizeof (struct GMT_XSEGMENT), gmtsupport_ysort, y);
                                                      ^~~~~~~~~~~~~~~~
/Users/pwessel/UH/RESEARCH/CVSPROJECTS/GMTdev/gmt-dev/src/gmt_dev.h:187:87: note: expanded from macro 'QSORT_R'
    #define QSORT_R(base, nel, width, compar, thunk) qsort_r(base, nel, width, thunk, compar);
                                                                                      ^~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/stdlib.h:345:22: note: passing argument to parameter '__compar' here
            int (* _Nonnull __compar)(void *, const void *, const void *));

```

Need to know what Windows does, @joa-quim.  Note the original problem seems limited to Apple Silicon (which I am on) so have not added `_arm_` stuff yet.

### CODE REVIEW

**Comment by PaulWessel** [comment]
Getting some warnings on the QSORT_R macro but 

`gmt surface xy_vert.txt -Gvert.grd -R-125.8/-122.6/39.3/41.7 -I0.0064 -rp`

no longer crash for me.  Please check this branch out, @kmaterna.

**Comment by PaulWessel** [comment]
I am getting many failures involving **surface** so definitively changes stuff.

**Comment by PaulWessel** [comment]
@joa-quim, what is the Windows _qsort_r_ man page say about the order of the arguments?

**Comment by joa-quim** [comment]
Forgot all about this. I only find ``qsort_s`` not ``qsort_r`` in MSDN
https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/qsort-s?view=msvc-170

**Comment by kmaterna** [comment]
I was able to check the branch out on M1 and we can verify that the segmentation fault no longer happens.  however, the written grd file under -rg and -rp is not the expected output.  
[produced_grds.zip](https://github.com/GenericMappingTools/gmt/files/11944025/produced_grds.zip)

**Comment by PaulWessel** [comment]
Yep. We are getting lots of failures in our tests as well.

**Comment by PaulWessel** [comment]
Ah, wouldn't ou know it.  The elected order of the 3 argument to the comparison function is also different.  For macOS the first argument is the fixed pointer, while for Linux it is the 3rd argument that is the fixed pointer.
I have made that change under suitable ifdefs and now I think we are getting somewhere: My lots of new failures reported above are gone and no crash on the original surface command.  Please check. @joa-quim I also added a macro so we can use qsort_s on Windows - it has the same pointer order as macOS. So consider still this PR as WIP and let me know if all goes well.  You should get a (temporary) warning at the start of the compile (when it builds compatibility/qsort.c that says it will use qsort_s instead.

File: src/compat/qsort.c [hunk 1]
```diff
@@ -255,3 +258,10 @@ _quicksort (void *const pbase, size_t total_elems, size_t size,
   }
 }
 #endif /* !HAVE_QSORT_R_GLIBC */
+#else
+#if _WIN32
```
Comment (joa-quim) [inline src/compat/qsort.c:262]: If the intention is to target the VS compiler then the macro to use must be _MSC_VER. _WIN32 affects also the MinGW compiler (gcc with I don't know whats mixes)

**Comment by joa-quim** [comment]
But also got compiling errors and warnings
```
C:\progs_cygw\GMTdev\gmt5\master\src\compat\qsort.c(263): fatal error C1021: invalid preprocessor command 'warning'
[61/566] Building C object src\CMakeFiles\gmtlib.dir\gmt_support.c.obj
C:\progs_cygw\GMTdev\gmt5\master\src\gmt_support.c(14851): warning C4047: 'function': '_CoreCrtSecureSearchSortCompareFunction' differs in levels of indirection from 'double *'
C:\progs_cygw\GMTdev\gmt5\master\src\gmt_support.c(14851): warning C4024: 'qsort_s': different types for formal and actual parameter 4
```

File: src/compat/qsort.c (hunk 1 above)
Comment (PaulWessel) [reply src/compat/qsort.c]: OK, fixed

**Comment by PaulWessel** [comment]
Did you get similar warnings from surface.c?  Not sure why since the docs says

```
void qsort_s(
   void *base,
   size_t num,
   size_t width,
   int (__cdecl *compare )(void *, const void *, const void *),
   void * context
);

```
so the order of args and args in compare are the same as for macOS.

**Comment by joa-quim** [comment]
It doesn't advance much more after
```
C:\progs_cygw\GMTdev\gmt5\master\src\compat\qsort.c(263): fatal error C1021: invalid preprocessor command 'warning'
```

**Comment by joa-quim** [comment]
Then there are more warnings of the same type

```
C:\progs_cygw\GMTdev\gmt5\master\src\surface.c(539): warning C4047: 'function': '_CoreCrtSecureSearchSortCompareFunction' differs in levels of indirection from 'SURFACE_SEARCH *'
C:\progs_cygw\GMTdev\gmt5\master\src\surface.c(539): warning C4024: 'qsort_s': different types for formal and actual parameter 4
C:\progs_cygw\GMTdev\gmt5\master\src\surface.c(1312): warning C4047: 'function': '_CoreCrtSecureSearchSortCompareFunction' differs in levels of indirection from 'SURFACE_SEARCH *'
C:\progs_cygw\GMTdev\gmt5\master\src\surface.c(1312): warning C4024: 'qsort_s': different types for formal and actual parameter 4
C:\progs_cygw\GMTdev\gmt5\master\src\surface.c(1331): warning C4047: 'function': '_CoreCrtSecureSearchSortCompareFunction' differs in levels of indirection from 'SURFACE_SEARCH *'
C:\progs_cygw\GMTdev\gmt5\master\src\surface.c(1331): warning C4024: 'qsort_s': different types for formal and actual parameter 4
```

File: src/compat/qsort.c [hunk 2]
```diff
@@ -255,3 +258,10 @@ _quicksort (void *const pbase, size_t total_elems, size_t size,
   }
 }
 #endif /* !HAVE_QSORT_R_GLIBC */
+#else
+#if _MSC_VER
+#warning "C Preprocessor determined we will use system function qsort_s]"
```
Comment (joa-quim) [inline src/compat/qsort.c:263]: Replacing by
```
#pragma message("C Preprocessor determined we will use system function qsort_s]")
```
at least makes the compiling go silently through it.

Comment (PaulWessel) [reply src/compat/qsort.c]: OK, now using pragma message

**Comment by PaulWessel** [comment]
Try this latest - I had the order in qsort_s backwards...

File: src/compat/qsort.c [hunk 3]
```diff
@@ -260,8 +260,8 @@ _quicksort (void *const pbase, size_t total_elems, size_t size,
 #endif /* !HAVE_QSORT_R_GLIBC */
 #else
 #if _MSC_VER
```
Comment (joa-quim) [inline src/compat/qsort.c:262]: But now no need to use a ``#if _MSC_VER `` since both branches do the same thing.

Comment (PaulWessel) [reply src/compat/qsort.c:262]: True.  I cannot test this so best if you mess around until it is OK. Just made a change so hopefully working.

**Comment by joa-quim** [comment]
OK, run the tests and all seems fine.

**Comment by PaulWessel** [comment]
Hi @kmaterna, please try out this branch so hopefully I can remove the WIP.

**Comment by kmaterna** [comment]
This works now for me!

**Comment by PaulWessel** [comment]
@seisman, any comments on this PR? I think it works on Windows, macOS and Linux.  I dont know if there are Solaris or HP-UX computers anymore running actual UNIX, but it seems OK.  WIth this PR I am not sure if the combat dir serves any function anymore, but we can leave it in there for now since at least the ifdefs ensure it is not included unless ew cannot find _qsort_r_.

**Comment by seisman** [comment]
> @seisman, any comments on this PR? I think it works on Windows, macOS and Linux. I dont know if there are Solaris or HP-UX computers anymore running actual UNIX, but it seems OK. WIth this PR I am not sure if the combat dir serves any function anymore, but we can leave it in there for now since at least the ifdefs ensure it is not included unless ew cannot find _qsort_r_.

Looks good to me but I don't think I have enough knowledge to comment on it.

**Comment by PaulWessel** [comment]
OK, awaiting _qsort_r_ documentation for UNIX (e.g., Solaris) from SOET IT before we finalise this PR.

**Comment by PaulWessel** [comment]
Solaris shows

```
void qsort_r(void *base, size_t nel, size_t width,
    int (*compar)(const void *x, const void *y, void *context),
    void *context);

```
which is the same as macOS.  Please approve this one and we will merge it.

**Review by joa-quim — APPROVED** [review]

**Comment by joa-quim** [comment]
I'm afraid this is not over. I'm now having crashes in Julia tests in a gmtspatial call but the crash is in gmt_support#4836

```
#ifdef QSORT_R_THUNK_FIRST
/* arg is first argument to compare function */
GMT_LOCAL int gmtsupport_ysort (void *arg, const void *p1, const void *p2) {
#else
/* arg is last argument to compare function */
GMT_LOCAL int gmtsupport_ysort (const void *p1, const void *p2, void *arg) {
#endif
	const struct GMT_XSEGMENT *a = p1, *b = p2;
	double *x2sys_y = arg;

	if (x2sys_y[a->start] < x2sys_y[b->start]) return -1;
```

**Pull request merged**